### ISSUE TITLE
刚才理解错了，又去看了下论文，里面有个 old policy model 

### ISSUE THREAD

**Issue by QingQingS** [lsdefine/simple_GRPO#8]
你好，这两天仔细看了下您开源的代码，有些疑问请教下：
在读代码时发现policy model 对相同的question 进行了两次forward：
      1. generate_mode函数中计算reward时forward了一次 ；
      2.GRPO_step函数中计算per_token_logps时forward了一次     
为什么要进行两次forward呢，这样会浪费计算资源吧
相同的question两次foward 每次forward的生成的token是一致的么？

**Comment by QingQingS** [comment]
old policy model  的参数 在哪里更新的呢，代码中没看到对应的更新操作

<img width="437" alt="Image" src="https://github.com/user-attachments/assets/1df90766-e5bb-4ff9-ac0e-78884b7e0f73" />

**Comment by ZKCCZ** [comment]
> old policy model 的参数 在哪里更新的呢，代码中没看到对应的更新操作
> 
> <img alt="Image" width="437" src="https://private-user-images.githubusercontent.com/150359873/414591711-1df90766-e5bb-4ff9-ac0e-78884b7e0f73.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5NTE0MzYsIm5iZiI6MTczOTk1MTEzNiwicGF0aCI6Ii8xNTAzNTk4NzMvNDE0NTkxNzExLTFkZjkwNzY2LWU1YmItNGZmOS1hYzBlLTc4ODg0YjdlMGY3My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxOVQwNzQ1MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZTAyZDA3NzM4MWZjZWNiYmI0ODE4MTQxNDFhZjI5NDkxNjdhYWM2MzVlNDkzNzFiMGJkYWZmZWY0MjBmMDc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cBGy41NBCTc-6G_Yvbzr_Cxrow0rnkn-7o87uKRgdH4">

我看到gen_model = engine， 指向同一个模型，应该backward后会一起更新

**Comment by QingQingS** [comment]
那这样的话 ,两个模型就是一样的了吧，没有old 不old的区分了

**Comment by lsdefine** [comment]
如果保证用于训练的都是刚生成的答案，那么确实可以省掉一步推理
但是模型自生成自训练在速度和效果上都是有改进空间的，后面会在生成的时候加入大量干预诱导它生成更好的答案，那么这个时候肯定训练的时候要推理一次的

**Comment by chenlinzhe** [comment]
两次:生成模式：输入是问题提示（prompts），输出是多样化的候选答案（answers），用于探索和奖励计算，不涉及训练。
训练模式：输入是完整的序列（batch['inputs']），包含提示和生成的答案，输出是 logits 和 per_token_logps，用于训练和优化。

可:
在 generate_mode 中生成答案时，同时缓存 logits 或 per_token_logps，在 GRPO_step 中复用这些结果，减少重复 forward pass。
修改 gen_samples 返回 logits，并在 GRPO_step 中直接使用

**Comment by LYCnight** [comment]
> 两次:生成模式：输入是问题提示（prompts），输出是多样化的候选答案（answers），用于探索和奖励计算，不涉及训练。 训练模式：输入是完整的序列（batch['inputs']），包含提示和生成的答案，输出是 logits 和 per_token_logps，用于训练和优化。
> 
> 可: 在 generate_mode 中生成答案时，同时缓存 logits 或 per_token_logps，在 GRPO_step 中复用这些结果，减少重复 forward pass。 修改 gen_samples 返回 logits，并在 GRPO_step 中直接使用

训练模式：输入是完整的序列（batch['inputs']），包含提示和生成的答案，输出是 logits 和 per_token_logps，用于训练和优化。
这里我有一些不明白。这里的 inputs = prompt + question + answer, 这样输入后得到的输出是什么呢？是模型继续生成文字吗？

**Comment by Und3r1ine** [comment]
> > 两次:生成模式：输入是问题提示（prompts），输出是多样化的候选答案（answers），用于探索和奖励计算，不涉及训练。 训练模式：输入是完整的序列（batch['inputs']），包含提示和生成的答案，输出是 logits 和 per_token_logps，用于训练和优化。
> > 可: 在 generate_mode 中生成答案时，同时缓存 logits 或 per_token_logps，在 GRPO_step 中复用这些结果，减少重复 forward pass。 修改 gen_samples 返回 logits，并在 GRPO_step 中直接使用
> 
> 训练模式：输入是完整的序列（batch['inputs']），包含提示和生成的答案，输出是 logits 和 per_token_logps，用于训练和优化。 这里我有一些不明白。这里的 inputs = prompt + question + answer, 这样输入后得到的输出是什么呢？是模型继续生成文字吗？

我理解是通过最小化损失函数来训练模型，而损失函数的计算需要生成的结果和参考结果，因此输出应该是一段模型生成的答案。
